### PR TITLE
[3.3.5] Core/Pet: Voidwalker Spell 'Consume Shadows'

### LINKED ISSUE #14054: Core/Pet: Voidwalker - Interrupts Spell: 'Consume Shadows' while moving
npc=1860
http://www.wowhead.com/spell=36472/

The Voidwalker interrupts alway his cast 'Consume Shadows' if it is currently moving/following it's owner.

Demonstration:
http://youtu.be/rbRdcYri5Wg

How it should be:
Usually it should stay on it's place until it is finished with it's cannelcast.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8359796-core-pet-voidwalker-interrupts-spell-consume-shadows-while-moving?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ghost**
by @robinsch; closes  #14054

Voidwalker should stop moving until it has completed channeling 'Consume Shadows' .

Existing issue: if the Voidwalker moves/follows its owner,
it interrupts itself from casting 'Consume Shadows' :
http://www.wowhead.com/spell=36472/consume-shadows

### CODE REVIEW

**Comment by Nyeriah** [comment]
How does this deal with creatures that should channel spells while moving?

http://www.wowhead.com/spell=161199/debilitating-fixation

**Comment by robinsch** [comment]
You are correct.

```diff
From f8bc75bcbb44c79e0b0e69b6887fc6f75c08eec9 Mon Sep 17 00:00:00 2001
From: robinsch <robin.schriever.hude@web.de>
Date: Sun, 31 May 2015 17:14:29 +0200
Subject: [PATCH] Implement SPELL_ATTR_EX5_CAN_CHANNEL_WHEN_MOVING. This will
 not interrupt certain channeling spell casts for moving creatures. By @kvipka

---
 src/server/game/AI/SmartScripts/SmartScript.cpp | 5 ++++-
 src/server/game/Miscellaneous/SharedDefines.h   | 2 +-
 src/server/game/Spells/Spell.cpp                | 4 ++--
 3 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/server/game/AI/SmartScripts/SmartScript.cpp b/src/server/game/AI/SmartScripts/SmartScript.cpp
index 2ea7ed2..b995fb3 100644
--- a/src/server/game/AI/SmartScripts/SmartScript.cpp
+++ b/src/server/game/AI/SmartScripts/SmartScript.cpp
@@ -483,13 +483,13 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
                         if (e.action.cast.flags & SMARTCAST_INTERRUPT_PREVIOUS)
                             me->InterruptNonMeleeSpells(false);
 
+                        SpellInfo const* spellInfo = sSpellMgr->EnsureSpellInfo(e.action.cast.spell);
                         if (e.action.cast.flags & SMARTCAST_COMBAT_MOVE)
                         {
                             // If cast flag SMARTCAST_COMBAT_MOVE is set combat movement will not be allowed
                             // unless target is outside spell range, out of mana, or LOS.
 
                             bool _allowMove = false;
-                            SpellInfo const* spellInfo = sSpellMgr->EnsureSpellInfo(e.action.cast.spell);
                             int32 mana = me->GetPower(POWER_MANA);
 
                             if (me->GetDistance(*itr) > spellInfo->GetMaxRange(true) ||
@@ -501,6 +501,9 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
                             ENSURE_AI(SmartAI, me->AI())->SetCombatMove(_allowMove);
                         }
 
+                        if ((spellInfo->IsChanneled() || spellInfo->CalcCastTime()) && me->isMoving() && spellInfo->InterruptFlags & SPELL_INTERRUPT_FLAG_MOVEMENT && !spellInfo->HasAttribute(SPELL_ATTR5_CAN_CHANNEL_WHILE_MOVING))
+                            me->StopMoving();
+
                         me->CastSpell((*itr)->ToUnit(), e.action.cast.spell, (e.action.cast.flags & SMARTCAST_TRIGGERED) != 0);
                     }
                     else if (go)
diff --git a/src/server/game/Miscellaneous/SharedDefines.h b/src/server/game/Miscellaneous/SharedDefines.h
index 06c58a4..251c83d 100644
--- a/src/server/game/Miscellaneous/SharedDefines.h
+++ b/src/server/game/Miscellaneous/SharedDefines.h
@@ -456,7 +456,7 @@ enum SpellAttr4
 
 enum SpellAttr5
 {
-    SPELL_ATTR5_UNK0                             = 0x00000001, //  0
+    SPELL_ATTR5_CAN_CHANNEL_WHILE_MOVING         = 0x00000001, //  0 don't interrupt channeling spells when moving
     SPELL_ATTR5_NO_REAGENT_WHILE_PREP            = 0x00000002, //  1 not need reagents if UNIT_FLAG_PREPARATION
     SPELL_ATTR5_UNK2                             = 0x00000004, //  2
     SPELL_ATTR5_USABLE_WHILE_STUNNED             = 0x00000008, //  3 usable while stunned
diff --git a/src/server/game/Spells/Spell.cpp b/src/server/game/Spells/Spell.cpp
index be0501e..3c78540 100644
--- a/src/server/game/Spells/Spell.cpp
+++ b/src/server/game/Spells/Spell.cpp
@@ -2927,7 +2927,7 @@ void Spell::prepare(SpellCastTargets const* targets, AuraEffect const* triggered
 
     // don't allow channeled spells / spells with cast time to be cast while moving
     // (even if they are interrupted on moving, spells with almost immediate effect get to have their effect processed before movement interrupter kicks in)
-    if ((m_spellInfo->IsChanneled() || m_casttime) && m_caster->GetTypeId() == TYPEID_PLAYER && m_caster->isMoving() && m_spellInfo->InterruptFlags & SPELL_INTERRUPT_FLAG_MOVEMENT)
+    if (((m_spellInfo->IsChanneled() && !m_spellInfo->HasAttribute(SPELL_ATTR5_CAN_CHANNEL_WHILE_MOVING)) || m_casttime) && m_caster->GetTypeId() == TYPEID_PLAYER && m_caster->isMoving() && m_spellInfo->InterruptFlags & SPELL_INTERRUPT_FLAG_MOVEMENT)
     {
         SendCastResult(SPELL_FAILED_MOVING);
         finish(false);
@@ -3262,7 +3262,7 @@ void Spell::handle_immediate()
         }
 
         // interrupt movement at channeled spells for creature case
-        if (m_caster->GetTypeId() == TYPEID_UNIT && m_caster->isMoving())
+        if (m_caster->GetTypeId() == TYPEID_UNIT && m_caster->isMoving() && !m_spellInfo->HasAttribute(SPELL_ATTR5_CAN_CHANNEL_WHILE_MOVING))
             m_caster->StopMoving();
     }
 
-- 
1.9.5.msysgit.1

**Comment by ShinDarth** [comment]
Rev: 302623a1516e487ad64f2028aec287b99372e33a

Current behavior (on clean TrinityCore):

- when Voidwalker is moving, you are unable to cast Consume Shadows
- when is not moving, you can cast Consume Shadows and the Voidwalker will **not** follow you (unless you manually tell him to follow you)
- when Voidwalker is casting Consume Shadows and you click on follow button, the Voidwalker will interrupt casting and will follow you

**Comment by tkrokli** [comment]
Thank you for confirming the issue on recent source. Have you tested if this patch works for you? :)

**Comment by Kittnz** [comment]
Gonna test tonight, i think this might fix a few other bugs aswell

**Comment by MitchesD** [comment]
any news here @Kittnz ?

File: src/server/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -3271,6 +3271,10 @@ void Spell::handle_immediate()
             m_caster->AddInterruptMask(m_spellInfo->ChannelInterruptFlags);
             SendChannelStart(duration);
         }
+
```
Comment (Runico) [inline src/server/game/Spells/Spell.cpp]: whitespaces

**Comment by Kittnz** [comment]
turns out it didn't fix the issue i was after, i didn't test the voidwalker yet

File: src/server/game/Spells/Spell.cpp (hunk 1 above)
Comment (tkrokli) [inline src/server/game/Spells/Spell.cpp]: OK, will fix.

**Comment by jackpoz** [comment]
what's the current status of this PR ?

**Comment by tkrokli** [comment]
Except for keeping it in sync with the TC source, I don't know. I hope @robinsch can tell you more.
Because it was made as a response to @Rushor's issue  #14054, I thought he would be interested.

**Comment by robinsch** [comment]
same status as before, finished

**Comment by jackpoz** [comment]
If you run, let voidwalker follow you and click on Consume Shadows it will cast it for a moment and then interrupt it and keep following you

**Comment by ShinDarth** [comment]
@jacpoz I was wondering the same thing, but I guess it isn't intended.

**Comment by tkrokli** [comment]
@robinsch : any idea if the thing jackpoz and ShinDarth mention is something that can be fixed here?

**Comment by robinsch** [comment]
No this is not intended as it was how it worked before. 7m old pr so idk if something changed that make this no longer work.

**Comment by Keader** [comment]
I'm using the fix from the PR was created and always was what Jack said

**Comment by tkrokli** [comment]
Care to continue working on this PR, @robinsch ? Or do you want to have it closed?

**Comment by ChazyTheBest** [comment]
I can confirm that (in tbc/wotlk retail) If you are running and click Consume Shadows, the Voidwalker will stop moving and begin the casting. And he won't follow you until he finish (unless you click follow or attack to a near mob). I remember it perfectly.

**Comment by tkrokli** [comment]
Oh well. We will be needing a new PR, then.

**Comment by ChazyTheBest** [comment]
Sorry for being late :P

**Comment by ccrs** [comment]
this is correct, the problem is that pet follow movement cancels the channeling.

**Comment by ShinDarth** [comment]
merge?

**Comment by Keader** [comment]
@ShinDarth still have problem

**Comment by tkrokli** [comment]
@ShinDarth : it appears that this PR does not solve the issue it was supposed to solve.

**Comment by ShinDarth** [comment]
the why is it still open?

**Comment by tkrokli** [comment]
Well, I did not consider the option. Usually it is the TC devs who decide if it should be closed or not.

**Comment by Keader** [comment]
Needs improvement, but the fix works. (Work if he dont move on start cast)
Just does not work when during attempts to cast while moving. You only need to fix it

**Comment by ShinDarth** [comment]
@Keader isn't that the default behaviour on current TC (without this PR) ?

**Comment by Keader** [comment]
Nop, without the PR to get cast (with  void stopped) and the player to move, the cast is cut to follow the owner.

**Comment by ccrs** [comment]
https://github.com/TrinityCore/TrinityCore/pull/14813#issuecomment-169377104

If u want to solve the issue, add something like a check in petai update

**Comment by ShinDarth** [comment]
If this PR actually solves something, and if it does it in the proper way, I think it can be merged even if there are still other things that needs to be improved... Otherwise it should be closed.

**Comment by Kittnz** [comment]
i haven't found anything not working since i tested this pr (just local build)

**Comment by jackpoz** [comment]
@Kittnz what about https://github.com/TrinityCore/TrinityCore/pull/14813#issuecomment-166011194 ?

**Comment by ccrs** [comment]
the attribute is implemented correctly, its just the movement calls for each AI the ones that need to be revised (unrelated to the attrubute itself).

**Comment by ChazyTheBest** [comment]
So the only remaining problem is: when you're moving and click the spell not let the pet cancel it and follow you until he finishes the channeling (everything else should cancel it).

**Comment by Keader** [comment]
True, last time I tested the TC tava one month out of date, I thought it would make no difference (I had not seen any commits related to that) but aparemente did.
Thanks @ChazyTheBest

**Comment by Kittnz** [comment]
So this works perfectly without the PR so no need anymore for this PR?

**Comment by Keader** [comment]
@Kittnz Dont need anymore

**Comment by ShinDarth** [comment]
ok, closing.

**Comment by tkrokli** [comment]
OK, so at least I got to clean off one more branch from my repo. :-p

**Comment by Kittnz** [comment]
Great, another closed PR, up to the next :+1:

**Comment by ChazyTheBest** [comment]
https://github.com/TrinityCore/TrinityCore/pull/14813#issue-82978518 The original problem is still there...

**Comment by tkrokli** [comment]
It needs to be fixed by someone who knows how to handle it in the current source.
When I made this PR, I did it as a service to robinsch who could not make TC Pull Requests in his repo.

**Pull request closed without merging**